### ISSUE TITLE
How to add anything on powerline (spaceline) ?

### ISSUE THREAD

**Issue by ZaenChou** [syl20bnr/spacemacs#9458]
So I am trying to get 2 stuff working,

* displaying date-time on powerline (spaceline)
* spaceline-all-the-icons from here: https://github.com/domtronn/spaceline-all-the-icons.el

but the documentation is outdated\* i guess, and the 10-hour trial and error gave me no luck at all.
   \* The part I am referring to is here: 
   https://github.com/syl20bnr/spacemacs/blob/master/doc/DOCUMENTATION.org#customizing-the-mode-line
   the definition is void, spacemacs says
   and the spaceline documentation is here: https://github.com/TheBB/spaceline#segments
   this one is new (and accurate?), I tried this as well, still no luck

#### Issues:
1. Which is it, powerline or spaceline? 😕 As I understand it, spacemacs calls it powerline, the spacemacs independent https://github.com/TheBB/spaceline calls it spaceline. But there are both packages in spacemacs according to package-selected-packages, so, WUH? And this leads to the confusion of which package should I wrap my code around.
2. According to spaceline documentation, the correct way of defining segment is "spaceline-define-segment", but is that it? there's no right or left property mentioned (there are tight-left and tight-right though, no idea what it means), and after defining it, what?
3. Has anyone tried to make spaceline-all-the-icons stuff working on spacemacs?

maybe it is not most fitting to post this on spacemacs repo, but again, I'm not sure about the whole naming thing, and TheBB is as everyone knows one of the spacemacs contributors, so I decided to post here anyway

##### date-time
1. simply put it in user-config()
1. put it in user-init()
1. (spacemacs|use-package-add-hook) it
 1 pre-config
 2 .. you now how it goes
1. realize the naming thing, try again with powerline
##### icons
the tricky thing: :after spaceline
1. user-config()
1. user-init()
1. (spacemacs|use-package-add-hook)
  1  those things without :after spaceline
  2  with
1. wrap (spacemacs|use-package-add-hook) with eval-after-load

my .spacemacs.d/init.el is here: https://github.com/ZaenChou/dotfiles-desktop/blob/master/.spacemacs.d/init.el

any help is appreciated 👍

**Comment by CeleritasCelery** [comment]
As you mentioned above, the best place for this question is [spaceline](https://github.com/TheBB/spaceline)

**Comment by ZaenChou** [comment]
@CeleritasCelery 
Alright, but I still don't understand which name is used in spacemacs, do you think I'd better close this and open another issue for it? Or just rename this issue and keep the thread?

**Comment by CeleritasCelery** [comment]
Spacemacs  powerline is called spaceline. If you want to learn how to customize it, read the docs at
https://github.com/TheBB/spaceline/blob/master/README.org#moderate-configuration

**Comment by jupl** [comment]
In Spacemacs for showing date/time there is the command `SPC t m t` or `(spacemacs/toggle-display-time-on)`.

**Comment by ZaenChou** [comment]
@jupl 
WOW I did not know there was a modeline toggling section in SPC t
Thanks a lot man! (I can even disable modeline itself what the hell)

**Comment by ZaenChou** [comment]
@jupl 
Sorry to bother you again, but is there any way to specify the time format? I want it to show the date and second as well, and in the 24-hour clock.

**Comment by duianto** [comment]
### tl;dr
Add these lines in the `.spacemacs` files `dotspacemacs/user-config` section:
```
  ;; mode line time stamp
  (setq display-time-24hr-format t)
  (setq display-time-format "%H:%M:%S")        ; add seconds
  (setq display-time-interval 1)               ; update every second
  (setq display-time-default-load-average nil) ; don't show load average
  (setq display-time-mail-string "")           ; don't show mail
  (display-time-mode 1)                 ; show time in mode line on startup
```

### The long way around:
This is how figured out how to set the mode line time stamp to 24h with seconds:
Show the description for the mode lines display-time command:
`SPC h d k SPC t m t`
Press `o` to call `ace-link-help` which shows a highlighted letter, next to each link in the help buffer.

If no previous help pages had been opened, then the only link will be `‘../../keybindings.el’`, and it will open automatically, otherwise a letter will be shown, most likely `a` since it's the first link.

If it didn't open the `keybindings.el` file automatically, then press `a` to open it.

`/` search for `display-time`, notice that it calls `display-time-mode`.
With the cursor on `display-time-mode`, press `gd` to go to it's definition.

`/` search for `24`. The first instance is the variable `display-time-24hr-format`.

It's doc-string states that:
> Non-nil indicates time should be displayed as hh:mm, 0 <= hh <= 23.

That means that the command we want is:
`(setq display-time-24hr-format t)`

### To show the seconds:
`/` search for `second`, there are both a `display-time-format` and `display-time-string`, I was unsure which to set, so a quick googling, found this: https://emacs.stackexchange.com/questions/6065/how-to-display-time-in-seconds-in-the-mode-line

It shows this command:
`(setq display-time-format "%I:%M:%S")`
But the doc-string for `format-time-string` shows that: 
`%I` is for a 12-hour clock and 
`%H` is for a 24-hour clock, therefore the command becomes:
`(setq display-time-format "%H:%M:%S")`

While browsing through the occurrences of `second`, I noticed this variable:
```
(defcustom display-time-interval 60
  "Seconds between updates of time in the mode line."
```

Setting it to:
`(setq display-time-interval 1)`
and calling `M-x display-time` starts updating the times tamp every second for this session. Adding it to the `.spacemacs` user-config section applies it on every startup.

### To show the time in the modeline on startup:
`SPC t m t` calls: `(spacemacs/toggle-display-time)` which in turn calls: `display-time-mode`, and it's doc-string states that:
> With a prefix argument ARG, enable Display Time mode if ARG is
positive, and disable it otherwise.  If called from Lisp, enable
it if ARG is omitted or nil.

Therefore the command we want to call is:
`(display-time-mode 1)`

### load average and mail
The doc-string for `display-time-mode` also states that it shows the load average and mail in the mode line.

These can also be turned off with:
`(setq display-time-default-load-average nil)`
source: https://www.reddit.com/r/emacs/comments/2ziinn/displaytimemode_but_not_system_load/

and mail:
`(setq display-time-mail-string "")`
source: https://unix.stackexchange.com/questions/120219/hide-mail-in-emacs-mode-line
Maybe the mail indicator only shows up if one has setup Emacs to check for mail.

Putting all of them together with comments, resulted in the lines at the top of this post.

**Comment by ZaenChou** [comment]
@duianto 
Much appreciation for sharing your fruit with me, I wasn't expecting this XD
I suppose the system load stuff is that weird decimal right next to the clock? Damn that thing annoyed me so much. So happy to see it's gone.
Was going to say "Hey I want the date too" but luckily I found this https://www.gnu.org/software/emacs/manual/html_node/elisp/Time-Parsing.html
so no need to bother you 👍 

I guess all of my concerns are addressed (giving up on all-the-icons), and I am closing this for good. Many thanks to celery, jupl and duianto!
